src/lib.rs

```diff
@@ -0,0 +1,389 @@
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! libtypec-rs is a library that aims to provide a generic interface
+//! abstracting all platform complexity for user space to develop tools for
+//! efficient USB-C port management and efficient diagnostic and debugging
+//! tools to debug of system issues around USB-C/USB PD topology.
+//!
+//! The data structures and interface APIs are based on USB Type-C® Connector
+//! System Software Interface (UCSI) Specification for the most part.
+
+// Note: this library is written in Rust, but one of its goals is to provide a C
+// FFI. Cbindgen is a tool that generates C bindings for Rust code. It is used
+// to generate a compatible header file. It does *not* understand namespacing,
+// since namespaces are not officially in the C language. This means that the
+// Rust types used throughout the library retain a (more verbose) full name in
+// spite of the module they are declared in.
+
+#![cfg_attr(feature = "backtrace", feature(error_generic_member_access))]
+
+use std::io::Cursor;
+
+use bitstream_io::LittleEndian;
+use pd::PdPdo;
+use ucsi::GetAlternateModesRecipient;
+use ucsi::GetPdoSourceCapabilitiesType;
+use ucsi::GetPdosSrcOrSink;
+use ucsi::PdMessage;
+use ucsi::PdMessageRecipient;
+use ucsi::PdMessageResponseType;
+use ucsi::UcsiAlternateMode;
+use ucsi::UcsiCableProperty;
+use ucsi::UcsiCapability;
+use ucsi::UcsiConnectorCapability;
+use ucsi::UcsiConnectorStatus;
+
+pub mod backends;
+pub mod pd;
+pub mod typec;
+pub mod ucsi;
+pub mod vdo;
+
+pub type BitWriter<'a> = bitstream_io::BitWriter<Cursor<&'a mut [u8]>, LittleEndian>;
+pub type BitReader<'a> = bitstream_io::BitReader<Cursor<&'a [u8]>, LittleEndian>;
+pub type Result<T> = std::result::Result<T, crate::Error>;
+
+/// A trait that abstracts the platform-specific backend.
+pub trait OsBackend {
+    fn capabilities(&mut self) -> Result<UcsiCapability>;
+
+    fn connector_capabilties(&mut self, connector_nr: usize) -> Result<UcsiConnectorCapability>;
+
+    fn alternate_modes(
+        &mut self,
+        recipient: GetAlternateModesRecipient,
+        connector_nr: usize,
+    ) -> Result<Vec<UcsiAlternateMode>>;
+
+    fn cable_properties(&mut self, connector_nr: usize) -> Result<UcsiCableProperty>;
+
+    fn connector_status(&mut self, connector_nr: usize) -> Result<UcsiConnectorStatus>;
+
+    fn pd_message(
+        &mut self,
+        connector_nr: usize,
+        recipient: PdMessageRecipient,
+        response_type: PdMessageResponseType,
+    ) -> Result<PdMessage>;
+
+    #[allow(clippy::too_many_arguments)]
+    fn pdos(
```

Comment:
These extremely long method calls might be better served by taking arguments as a struct, so that you can actually differentiate the numbers at a glance? (It's admittedly a fair bit more annoying to pass but not *too* bad, wgpu does this even for the C/++ API.)